enhancements/network/on-prem-bgp-networking.md

```diff
@@ -0,0 +1,444 @@
+
+### Risks and Mitigations
+
+- This design heavily relies on the infrastructure network that is not controlled by OpenShift. Documentation and validations will be provided
+  to ensure that prerequisites are met before a deployment can be done using BGP
+- Performance can be degraded if infrastructure network has routing problems, but routing issues would also involve a general outage, not just control plane VIPs
+- The opinionated configuration for FRR might be too simple for now, but we'll adjust it as we get feedback from customers and field deployments
+
+### Drawbacks
+
+The proposed solution heavily relies on the network infrastructure at the data center level and if something goes wrong
+with BGP, the OpenShift cluster can become very unstable.
+Also, the solution remains complex to implement and corner cases will have to be documented and covered by testing when possible (e.g. day 2 operations).
+
+## Design Details
+
+### Test Plan
+
+Testing will be implemented for OpenStack at first and re-usable by other platforms.
+We'll provide a set of scripts that can install ToR routers and this will be used to replicate
+the network infrastructure that customers could run.
+The jobs will do the following:
+- Create a proxy node that runs FRR and will be the upstream router for the OpenShift cluster
+- Deploy the OpenShift cluster and provides an `install-config.yaml` that has the needed options for BGP
+- Run the desired e2e test suite
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+These following requirements will apply first for OpenStack, and once this has proven to be stable, we'll turn this on for other platforms:
+* Failure Domains is supported by the platform
+* The code has been merged into various components (API, installer, MCO, baremetal-runtimecfg)
+* We have an image for FRR
+* Feature is documented
+* Feature is tested in CI
+
+Once these requirements are met, then we'll deliver this feature as Tech Preview. The current desired target is OCP 4.13.
+
+#### Tech Preview -> GA
+
+To go GA, this feature will have to be proven to be stable enough and tested by our QE.
+The current desired target is OCP 4.14.
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+If FRR configuration file has to change, we'll have a monitor container that will apply the new configuration and reload FRR daemons.
+
+### Version Skew Strategy
+
+### Operational Aspects of API Extensions
+
+#### Failure Modes
+
+A failure of BGP to peer routes that help to reach the VIPs can have a critical impact on cluster health.
+This is why we'll implement proper health checks and monitors, and also that we'll provide recommendations
+for customers who will deploy this architecture, for example router redundancy.
+
+#### Support Procedures
+
+It won't be possible to upgrade a cluster from Keepalived/HAproxy to BGP for now.
```

Comment:
Why can't you change between the two?

Reply:
I guess this is technically feasible, maybe I was being too strict in the initial writting and upgrades could be supported between a traditional to BGP networking.

Reply:
For now, I changed the wording saying that this is not possible *now* but might be revisited in the future. I'm convinced that an upgrade would bring a certain level of complexity that is not required at the moment.